src/parse.yy

```diff
@@ -2829,26 +2829,33 @@ lAddTemplateDeclaration(TemplateParms *templateParmList, DeclSpecs *ds, Declarat
         // FIXME: source position
         Error(decl->pos, "Illegal \"typedef\" provided with function template.");
         return;
-    } else {
-        if (decl->type == nullptr) {
-            Assert(m->errorCount > 0);
-            return;
-        }
+    }
+    // We can't support extern "C"/extern "SYCL" for templates because
+    // we need mangling information.
+    if (ds->storageClass == SC_EXTERN_C || ds->storageClass == SC_EXTERN_SYCL) {
+        Error(decl->pos, "Illegal linkage provided with function template.");
+        return;
+    }
 
```

Comment:
I think it make sense for clarity and readability to assert on possible `storageClass` values at this point. But if you prefer not to, it's not a problem.